### PR TITLE
Parallelize training onto a Multi-GPU machine

### PR DESCRIPTION
**Pull request by markus-hinsche**
**Motivation**: This is the ML iteration cycle

It should be as fast as possible, so we can make more iterations when developing a model

**Problem:** Our training is slow. Since we need to train depthmap model, we train from scratch. Therefore we need to train for days (e.g. 9.4 days for our current best height model)

**Solution:** Parallelize the training onto 4 GPUs in 1 machine (instead 1 GPU in 1 machine)

**Results**
[AzureML Experiment](https://ml.azure.com/experiments/id/4d4b47ee-b56c-44ca-8966-1a8d77201ece?wsid=/subscriptions/9b5bbfae-d5d1-4aae-a2ca-75159c0c887d/resourceGroups/cgm-ml-prod-we-rg/providers/Microsoft.MachineLearningServices/workspaces/cgm-ml-prod-we-azml&tid=3a27c573-ec1a-4734-9cd3-3208af51794b)

**Time measurements**

| Run       | Comment        | Epoch duration  | Average GPU utilization | 
| -------- |:-------------:| -----:| ----:|
| Run1      | BS=256, baseline                   | 822s | 24.5% |
| Run4     | BS=64, MirroredStrategy                  | 277s |  90.8%|
| Run10     | BS=256, MirroredStrategy                  | 221s | 95.0% |
| Run11     | BS=64, baseline                  |  837s | 24.3% |
| [Run12](https://ml.azure.com/experiments/id/4d4b47ee-b56c-44ca-8966-1a8d77201ece/runs/training-junkyard-multigpu_1625643816_d60a000f?wsid=/subscriptions/9b5bbfae-d5d1-4aae-a2ca-75159c0c887d/resourceGroups/cgm-ml-prod-we-rg/providers/Microsoft.MachineLearningServices/workspaces/cgm-ml-prod-we-azml&tid=3a27c573-ec1a-4734-9cd3-3208af51794b#details)     | BS=256, MirroredStrategy, Tf2.4 |  170s | 94.1% |
| [Run13](https://ml.azure.com/runs/training-junkyard-multigpu_1625657516_abf15b12?wsid=/subscriptions/9b5bbfae-d5d1-4aae-a2ca-75159c0c887d/resourcegroups/cgm-ml-prod-we-rg/workspaces/cgm-ml-prod-we-azml&tid=3a27c573-ec1a-4734-9cd3-3208af51794b#resourceMetrics)     | BS=512, MirroredStrategy                  |  164s |  93.7% |
| [Run14](https://ml.azure.com/runs/training-junkyard-multigpu_1625657560_4ee70fe5?wsid=/subscriptions/9b5bbfae-d5d1-4aae-a2ca-75159c0c887d/resourcegroups/cgm-ml-prod-we-rg/workspaces/cgm-ml-prod-we-azml&tid=3a27c573-ec1a-4734-9cd3-3208af51794b#resourceMetrics)     | BS=1024, MirroredStrategy                  |  174s |  92.6% |

**Improvements**
- **Speedup**: 3.0x to 3.7x (with TF 2.4 even 5x speedup) by having 4 GPUs instead of 1 (loss curve looks the same)
- We are able to use a  larger batch size (e.g. 1024 instead 256)

**Note:** 
- With TF2.3 sometimes we get a segmentation vault during the start (e.g. [Run9](https://ml.azure.com/experiments/id/4d4b47ee-b56c-44ca-8966-1a8d77201ece/runs/training-junkyard-multigpu_1624624557_86f9f67b?wsid=/subscriptions/9b5bbfae-d5d1-4aae-a2ca-75159c0c887d/resourcegroups/cgm-ml-prod-we-rg/workspaces/cgm-ml-prod-we-azml&tid=3a27c573-ec1a-4734-9cd3-3208af51794b)) -> let's keep this in mind and check at the beginning of the training if it was started successfully. Current problem: Random segfaults when using mirroredstrategy in/before the first epoch

### CODE REVIEW

File: src/models/CNNDepthMap/CNNDepthMap-height/q3-depthmap-plaincnn-height/train_notebook.ipynb [hunk 1]
```diff
@@ -56,8 +56,8 @@
    "outputs": [],
    "source": [
     "dataset_name = \"anon-depthmap-95k\"  # \"anon-depthmap-mini\"\n",
-    "experiment_name = \"q3-depthmap-plaincnn-height-95k\"\n",
-    "tags = {}"
+    "experiment_name = \"training-junkyard-multigpu\"\n",
+    "tags = {'run': 'mirrored strategy', 'TF': '2.4'}"
```
Comment (markus-hinsche) [inline src/models/CNNDepthMap/CNNDepthMap-height/q3-depthmap-plaincnn-height/train_notebook.ipynb:60]:
```suggestion
    "experiment_name = \"q3-depthmap-plaincnn-height-95k\"\n",
    "tags = {}"
```

**Review by gangooteli — APPROVED** [review]
LGTM

File: src/models/CNNDepthMap/CNNDepthMap-height/q3-depthmap-plaincnn-height/src/train.py [hunk 2]
```diff
@@ -154,52 +154,63 @@ def py_load_pickle(path, max_value):
...
+        dataset_training.batch(CONFIG.BATCH_SIZE),
+        validation_data=dataset_batches,
+        epochs=CONFIG.EPOCHS,
+        callbacks=training_callbacks,
+        verbose=2
+    )
+
+
+if CONFIG.USE_MULTIGPU:
+    strategy = tf.distribute.MirroredStrategy()
```
Comment (prajurock) [inline src/models/CNNDepthMap/CNNDepthMap-height/q3-depthmap-plaincnn-height/src/train.py:208]: `tf.distribute.MirroredStrategy()` What will be the output? Why strategy name?

**Review by prajurock — APPROVED** [review]

File: src/models/CNNDepthMap/CNNDepthMap-height/q3-depthmap-plaincnn-height/src/train.py (hunk 2 above)
Comment (markus-hinsche) [reply src/models/CNNDepthMap/CNNDepthMap-height/q3-depthmap-plaincnn-height/src/train.py:208]: > This strategy is typically used for training on one machine with multiple GPUs.

https://www.tensorflow.org/api_docs/python/tf/distribute/MirroredStrategy
